src/common/cmdparse.cc

```diff
@@ -155,16 +155,35 @@ cmdmap_from_json(vector<string> cmd, map<string, cmd_vartype> *mapp, stringstrea
       case json_spirit::array_type:
 	{
 	  // array is a vector of values.  Unpack it to a vector
-	  // of strings, the only type we handle.
+	  // of strings or int64_t, the only types we handle.
 	  vector<json_spirit::mValue> spvals = it->second.get_array();
-	  vector<string> outv;
-	  for (vector<json_spirit::mValue>::iterator sv = spvals.begin();
-	       sv != spvals.end(); ++sv) {
-	    if (sv->type() != json_spirit::str_type)
-	      throw(runtime_error("Can't handle arrays of non-strings"));
-	    outv.push_back(sv->get_str());
+	  vector<string> out_strs;
+	  vector<int64_t> out_ints;
+	  json_spirit::Value_type elem_type = json_spirit::null_type;
+	  for (const auto& sv : spvals) {
+	    if (elem_type == json_spirit::null_type) {
+	      elem_type = sv.type();
+	      if (elem_type != json_spirit::str_type &&
+		  elem_type != json_spirit::int_type) {
+		throw(runtime_error("Can't handle arrays of types other than "
+				    "int or string"));
+	      }
+	    }
+	    if (sv.type() != elem_type) {
+	      throw(runtime_error("Can't handle arrays of multiple types"));
+	    }
+	    if (elem_type == json_spirit::str_type)
+	      out_strs.push_back(sv.get_str());
+	    else
+	      out_ints.push_back(sv.get_int64());
 	  }
-	  (*mapp)[it->first] = outv;
+	  // if an empty array is acceptable, the caller should always check for
+	  // vector<string> if the expected value of "vector<int64_t>" in the
+	  // cmdmap is missing.
+	  if (out_ints.empty())
+	    (*mapp)[it->first] = out_strs;
+	  else
+	    (*mapp)[it->first] = out_ints;
 	}
```

Comment:
Would it not look better if you split look int 2 separate loops: for int64 and for strings?